sql/tests: TestRandomSyntaxFunctions failed

sql/tests.TestRandomSyntaxFunctions [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/11134564?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/11134564?buildTab=artifacts#/) on master @ [fcfa15d9694c0b840ac0c5c1b1489fa90a62efd8](https://github.com/cockroachdb/cockroach/commits/fcfa15d9694c0b840ac0c5c1b1489fa90a62efd8):

Fatal error:

```
panic: test timed out after 14m55s

```
Stack: 

```
goroutine 5187798 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x8e
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x32
```
<details><summary>Log preceding fatal error</summary>
<p>

```
    rsg_test.go:832: 45s of 5m0s: 144985 executions, 90074 successful
    rsg_test.go:832: 50s of 5m0s: 161358 executions, 100117 successful
    rsg_test.go:832: 55s of 5m0s: 178252 executions, 110757 successful
    rsg_test.go:832: 1m0s of 5m0s: 194554 executions, 120766 successful
    rsg_test.go:832: 1m5s of 5m0s: 206560 executions, 128234 successful
    rsg_test.go:832: 1m10s of 5m0s: 222997 executions, 138485 successful
    rsg_test.go:832: 1m15s of 5m0s: 239703 executions, 148821 successful
    rsg_test.go:832: 1m20s of 5m0s: 256356 executions, 159289 successful
    rsg_test.go:832: 1m25s of 5m0s: 273035 executions, 169458 successful
    rsg_test.go:832: 1m30s of 5m0s: 287272 executions, 178276 successful
    rsg_test.go:832: 1m35s of 5m0s: 304013 executions, 188607 successful
    rsg_test.go:832: 1m40s of 5m0s: 321278 executions, 199441 successful
    rsg_test.go:832: 1m45s of 5m0s: 338558 executions, 210168 successful
    rsg_test.go:832: 1m50s of 5m0s: 355607 executions, 220778 successful
    rsg_test.go:832: 1m55s of 5m0s: 372415 executions, 231119 successful
    rsg_test.go:832: 2m0s of 5m0s: 386533 executions, 239961 successful
    rsg_test.go:832: 2m5s of 5m0s: 403742 executions, 250553 successful
    rsg_test.go:832: 2m10s of 5m0s: 420319 executions, 260919 successful
    rsg_test.go:832: 2m15s of 5m0s: 436975 executions, 271183 successful
    rsg_test.go:832: 2m20s of 5m0s: 453789 executions, 281704 successful
    rsg_test.go:832: 2m25s of 5m0s: 470924 executions, 292259 successful
    rsg_test.go:832: 2m30s of 5m0s: 487276 executions, 302413 successful
    rsg_test.go:832: 2m35s of 5m0s: 502215 executions, 311581 successful
    rsg_test.go:832: 2m40s of 5m0s: 519277 executions, 322196 successful
    rsg_test.go:832: 2m45s of 5m0s: 536272 executions, 332689 successful
    rsg_test.go:832: 2m50s of 5m0s: 553185 executions, 343237 successful
    rsg_test.go:832: 2m55s of 5m0s: 570446 executions, 353916 successful
    rsg_test.go:832: 3m0s of 5m0s: 587236 executions, 364266 successful
    rsg_test.go:832: 3m5s of 5m0s: 600699 executions, 372581 successful
    rsg_test.go:832: 3m10s of 5m0s: 616980 executions, 382614 successful
    rsg_test.go:832: 3m15s of 5m0s: 634510 executions, 393535 successful
    rsg_test.go:832: 3m20s of 5m0s: 651195 executions, 403914 successful
    rsg_test.go:832: 3m25s of 5m0s: 668649 executions, 414773 successful
    rsg_test.go:832: 3m30s of 5m0s: 684841 executions, 424735 successful
    rsg_test.go:832: 3m35s of 5m0s: 700044 executions, 434308 successful
    rsg_test.go:832: 3m40s of 5m0s: 715905 executions, 444177 successful
    rsg_test.go:832: 3m45s of 5m0s: 733194 executions, 454960 successful
    rsg_test.go:832: 3m50s of 5m0s: 750034 executions, 465333 successful
    rsg_test.go:832: 3m55s of 5m0s: 767415 executions, 476100 successful
    rsg_test.go:832: 4m0s of 5m0s: 784547 executions, 486772 successful
    rsg_test.go:832: 4m5s of 5m0s: 799945 executions, 496392 successful
    rsg_test.go:832: 4m10s of 5m0s: 816287 executions, 506551 successful
    rsg_test.go:832: 4m15s of 5m0s: 832752 executions, 516803 successful
    rsg_test.go:832: 4m20s of 5m0s: 849077 executions, 526817 successful
    rsg_test.go:832: 4m25s of 5m0s: 865506 executions, 537018 successful
    rsg_test.go:832: 4m30s of 5m0s: 882249 executions, 547353 successful
    rsg_test.go:832: 4m35s of 5m0s: 899506 executions, 558164 successful
    rsg_test.go:832: 4m40s of 5m0s: 914070 executions, 567173 successful
    rsg_test.go:832: 4m45s of 5m0s: 931630 executions, 578053 successful
    rsg_test.go:832: 4m50s of 5m0s: 948722 executions, 588666 successful
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #107928 sql/tests: TestRandomSyntaxFunctions failed [C-test-failure O-robot T-sql-foundations branch-release-23.1]
- #95619 sql/tests: TestRandomSyntaxFunctions failed [C-test-failure O-robot T-sql-foundations branch-release-22.2]
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxFunctions.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-30260

sql/tests.TestRandomSyntaxFunctions [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/11150509?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/11150509?buildTab=artifacts#/) on master @ [32f846ff670bf26f1eb09309cdc7abae8878cf39](https://github.com/cockroachdb/cockroach/commits/32f846ff670bf26f1eb09309cdc7abae8878cf39):

Fatal error:

```
panic: test timed out after 14m55s

```
Stack: 

```
goroutine 4940619 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x8e
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x32
```
<details><summary>Log preceding fatal error</summary>
<p>

```
    rsg_test.go:832: 45s of 5m0s: 147127 executions, 91181 successful
    rsg_test.go:832: 50s of 5m0s: 164297 executions, 101904 successful
    rsg_test.go:832: 55s of 5m0s: 179583 executions, 111354 successful
    rsg_test.go:832: 1m0s of 5m0s: 195643 executions, 121377 successful
    rsg_test.go:832: 1m5s of 5m0s: 210157 executions, 130306 successful
    rsg_test.go:832: 1m10s of 5m0s: 226930 executions, 140812 successful
    rsg_test.go:832: 1m15s of 5m0s: 243849 executions, 151353 successful
    rsg_test.go:832: 1m20s of 5m0s: 261094 executions, 162117 successful
    rsg_test.go:832: 1m25s of 5m0s: 275966 executions, 171428 successful
    rsg_test.go:832: 1m30s of 5m0s: 292669 executions, 181704 successful
    rsg_test.go:832: 1m35s of 5m0s: 309461 executions, 192161 successful
    rsg_test.go:832: 1m40s of 5m0s: 326075 executions, 202409 successful
    rsg_test.go:832: 1m45s of 5m0s: 343405 executions, 213165 successful
    rsg_test.go:832: 1m50s of 5m0s: 360272 executions, 223692 successful
    rsg_test.go:832: 1m55s of 5m0s: 375238 executions, 232952 successful
    rsg_test.go:832: 2m0s of 5m0s: 392728 executions, 243794 successful
    rsg_test.go:832: 2m5s of 5m0s: 409288 executions, 254046 successful
    rsg_test.go:832: 2m10s of 5m0s: 426279 executions, 264717 successful
    rsg_test.go:832: 2m15s of 5m0s: 443143 executions, 275170 successful
    rsg_test.go:832: 2m20s of 5m0s: 459906 executions, 285667 successful
    rsg_test.go:832: 2m25s of 5m0s: 476010 executions, 295577 successful
    rsg_test.go:832: 2m30s of 5m0s: 491189 executions, 305160 successful
    rsg_test.go:832: 2m35s of 5m0s: 507638 executions, 315310 successful
    rsg_test.go:832: 2m40s of 5m0s: 524875 executions, 325999 successful
    rsg_test.go:832: 2m45s of 5m0s: 541911 executions, 336658 successful
    rsg_test.go:832: 2m50s of 5m0s: 557959 executions, 346453 successful
    rsg_test.go:832: 2m55s of 5m0s: 573004 executions, 355882 successful
    rsg_test.go:832: 3m0s of 5m0s: 590040 executions, 366378 successful
    rsg_test.go:832: 3m5s of 5m0s: 606379 executions, 376547 successful
    rsg_test.go:832: 3m10s of 5m0s: 622987 executions, 386804 successful
    rsg_test.go:832: 3m15s of 5m0s: 640043 executions, 397343 successful
    rsg_test.go:832: 3m20s of 5m0s: 655873 executions, 407137 successful
    rsg_test.go:832: 3m25s of 5m0s: 672585 executions, 417542 successful
    rsg_test.go:832: 3m30s of 5m0s: 687066 executions, 426488 successful
    rsg_test.go:832: 3m35s of 5m0s: 704355 executions, 437193 successful
    rsg_test.go:832: 3m40s of 5m0s: 720958 executions, 447540 successful
    rsg_test.go:832: 3m45s of 5m0s: 737347 executions, 457771 successful
    rsg_test.go:832: 3m50s of 5m0s: 754757 executions, 468650 successful
    rsg_test.go:832: 3m55s of 5m0s: 769363 executions, 477640 successful
    rsg_test.go:832: 4m0s of 5m0s: 786478 executions, 488180 successful
    rsg_test.go:832: 4m5s of 5m0s: 802814 executions, 498298 successful
    rsg_test.go:832: 4m10s of 5m0s: 819482 executions, 508650 successful
    rsg_test.go:832: 4m15s of 5m0s: 837066 executions, 519533 successful
    rsg_test.go:832: 4m20s of 5m0s: 852273 executions, 529011 successful
    rsg_test.go:832: 4m25s of 5m0s: 868611 executions, 539113 successful
    rsg_test.go:832: 4m30s of 5m0s: 884921 executions, 549215 successful
    rsg_test.go:832: 4m35s of 5m0s: 901652 executions, 559608 successful
    rsg_test.go:832: 4m40s of 5m0s: 918332 executions, 570021 successful
    rsg_test.go:832: 4m45s of 5m0s: 935628 executions, 580790 successful
    rsg_test.go:832: 4m50s of 5m0s: 952350 executions, 591106 successful
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #107928 sql/tests: TestRandomSyntaxFunctions failed [C-test-failure O-robot T-sql-foundations branch-release-23.1]
- #95619 sql/tests: TestRandomSyntaxFunctions failed [C-test-failure O-robot T-sql-foundations branch-release-22.2]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxFunctions.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/tests.TestRandomSyntaxFunctions [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/11169507?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/11169507?buildTab=artifacts#/) on master @ [4fe2a80d81c6fc5a3da3c7c44c5fc38da67e0367](https://github.com/cockroachdb/cockroach/commits/4fe2a80d81c6fc5a3da3c7c44c5fc38da67e0367):

Fatal error:

```
panic: test timed out after 14m55s

```
Stack: 

```
goroutine 5750783 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x8e
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x32
```
<details><summary>Log preceding fatal error</summary>
<p>

```
    rsg_test.go:832: 40s of 5m0s: 124135 executions, 76678 successful
    rsg_test.go:832: 45s of 5m0s: 140330 executions, 86766 successful
    rsg_test.go:832: 50s of 5m0s: 156005 executions, 96470 successful
    rsg_test.go:832: 55s of 5m0s: 172565 executions, 106805 successful
    rsg_test.go:832: 1m0s of 5m0s: 187321 executions, 115842 successful
    rsg_test.go:832: 1m5s of 5m0s: 198046 executions, 122535 successful
    rsg_test.go:832: 1m10s of 5m0s: 213663 executions, 132263 successful
    rsg_test.go:832: 1m15s of 5m0s: 230395 executions, 142542 successful
    rsg_test.go:832: 1m20s of 5m0s: 245905 executions, 152129 successful
    rsg_test.go:832: 1m25s of 5m0s: 262475 executions, 162453 successful
    rsg_test.go:832: 1m30s of 5m0s: 278229 executions, 172172 successful
    rsg_test.go:832: 1m35s of 5m0s: 291772 executions, 180564 successful
    rsg_test.go:832: 1m40s of 5m0s: 308090 executions, 190527 successful
    rsg_test.go:832: 1m45s of 5m0s: 323811 executions, 200254 successful
    rsg_test.go:832: 1m50s of 5m0s: 339742 executions, 210041 successful
    rsg_test.go:832: 1m55s of 5m0s: 356287 executions, 220341 successful
    rsg_test.go:832: 2m0s of 5m0s: 370934 executions, 229326 successful
    rsg_test.go:832: 2m5s of 5m0s: 386374 executions, 238889 successful
    rsg_test.go:832: 2m10s of 5m0s: 400232 executions, 247470 successful
    rsg_test.go:832: 2m15s of 5m0s: 416941 executions, 257854 successful
    rsg_test.go:832: 2m20s of 5m0s: 433411 executions, 268010 successful
    rsg_test.go:832: 2m25s of 5m0s: 449299 executions, 277819 successful
    rsg_test.go:832: 2m30s of 5m0s: 466189 executions, 288339 successful
    rsg_test.go:832: 2m35s of 5m0s: 482442 executions, 298358 successful
    rsg_test.go:832: 2m40s of 5m0s: 497602 executions, 307711 successful
    rsg_test.go:832: 2m45s of 5m0s: 514540 executions, 318235 successful
    rsg_test.go:832: 2m50s of 5m0s: 530748 executions, 328140 successful
    rsg_test.go:832: 2m55s of 5m0s: 547051 executions, 338270 successful
    rsg_test.go:832: 3m0s of 5m0s: 563025 executions, 348230 successful
    rsg_test.go:832: 3m5s of 5m0s: 579538 executions, 358361 successful
    rsg_test.go:832: 3m10s of 5m0s: 593225 executions, 366780 successful
    rsg_test.go:832: 3m15s of 5m0s: 609620 executions, 376951 successful
    rsg_test.go:832: 3m20s of 5m0s: 626223 executions, 387224 successful
    rsg_test.go:832: 3m25s of 5m0s: 641334 executions, 396679 successful
    rsg_test.go:832: 3m30s of 5m0s: 656457 executions, 406004 successful
    rsg_test.go:832: 3m35s of 5m0s: 672809 executions, 416167 successful
    rsg_test.go:832: 3m40s of 5m0s: 686547 executions, 424683 successful
    rsg_test.go:832: 3m45s of 5m0s: 702429 executions, 434422 successful
    rsg_test.go:832: 3m50s of 5m0s: 718869 executions, 444625 successful
    rsg_test.go:832: 3m55s of 5m0s: 735495 executions, 454943 successful
    rsg_test.go:832: 4m0s of 5m0s: 751777 executions, 465073 successful
    rsg_test.go:832: 4m5s of 5m0s: 767752 executions, 474865 successful
    rsg_test.go:832: 4m10s of 5m0s: 781263 executions, 483252 successful
    rsg_test.go:832: 4m15s of 5m0s: 797748 executions, 493538 successful
    rsg_test.go:832: 4m20s of 5m0s: 813606 executions, 503384 successful
    rsg_test.go:832: 4m25s of 5m0s: 829643 executions, 513208 successful
    rsg_test.go:832: 4m30s of 5m0s: 843978 executions, 522164 successful
    rsg_test.go:832: 4m35s of 5m0s: 860276 executions, 532318 successful
    rsg_test.go:832: 4m40s of 5m0s: 875368 executions, 541549 successful
    rsg_test.go:832: 4m45s of 5m0s: 890040 executions, 550589 successful
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #107928 sql/tests: TestRandomSyntaxFunctions failed [C-test-failure O-robot T-sql-foundations branch-release-23.1]
- #95619 sql/tests: TestRandomSyntaxFunctions failed [C-test-failure O-robot T-sql-foundations branch-release-22.2]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxFunctions.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/tests/rsg_test.go
+++ b/pkg/sql/tests/rsg_test.go
@@ -200,6 +200,16 @@ func (db *verifyFormatDB) execWithResettableTimeout(
 					return &nonCrasher{sql: sql, err: err}
 				}
 				return nil
+			case <-ctx.Done():
+				// Sanity: The context is cancelled when the test is about to
+				// timeout. We will log whatever statement we're waiting on for
+				// debugging purposes. Sometimes queries won't respect
+				// cancellation due to lib/pq limitations.
+				t.Logf("Context cancelled while executing: %q", sql)
+				// We will intentionally retry, which will us to wait for the
+				// go routine to complete above to avoid leaking it.
+				retry = true
+				return nil
 			case <-time.After(targetDuration):
 				db.mu.Lock()
 				defer db.mu.Unlock()
@@ -357,7 +367,9 @@ func TestRandomSyntaxFunctions(t *testing.T) {
 				case "crdb_internal.reset_sql_stats",
 					"crdb_internal.check_consistency",
 					"crdb_internal.request_statement_bundle",
-					"crdb_internal.reset_activity_tables":
+					"crdb_internal.reset_activity_tables",
+					"crdb_internal.revalidate_unique_constraints_in_all_tables",
+					"crdb_internal.validate_ttl_scheduled_jobs":
 					// Skipped due to long execution time.
 					continue
 				}

